src/qcodes/parameters/infer.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import TYPE_CHECKING, ClassVar
+
+from qcodes.instrument import Instrument, InstrumentBase, InstrumentModule
+from qcodes.instrument.parameter import DelegateParameter, Parameter
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+DOES_NOT_EXIST = "Does not exist"
+
+
+class InferError(AttributeError): ...
+
+
+class InferAttrs:
+    """Holds a global set of attribute name that will be inferred"""
+
+    _known_attrs: ClassVar[set[str]] = set()
+
+    @classmethod
+    def add_attrs(cls, attrs: str | Iterable[str]) -> None:
+        if isinstance(attrs, str):
+            attrs = (attrs,)
+        cls._known_attrs.update(set(attrs))
+
+    @classmethod
+    def known_attrs(cls) -> tuple[str, ...]:
+        return tuple(cls._known_attrs)
+
+    @classmethod
+    def discard_attr(cls, attr: str) -> None:
+        cls._known_attrs.discard(attr)
+
+    @classmethod
+    def clear_attrs(cls) -> None:
+        cls._known_attrs = set()
```

Comment:
Suggesting less verbose names because these methods are called on a class that’s called attrs so it’s clear what they add/discard/etc
```suggestion
class InferAttrs:
    """Holds a global set of attribute name that will be inferred"""

    _known_attrs: ClassVar[set[str]] = set()

    @classmethod
    def add(cls, attrs: str | Iterable[str]) -> None:
        if isinstance(attrs, str):
            attrs = (attrs,)
        cls._known_attrs.update(set(attrs))

    @classmethod
    def known_attrs(cls) -> tuple[str, ...]:
        return tuple(cls._known_attrs)

    @classmethod
    def discard(cls, attr: str) -> None:
        cls._known_attrs.discard(attr)

    @classmethod
    def clear(cls) -> None:
        cls._known_attrs = set()
```